Sources/MongoSwift/BSON/Codec.swift

```diff
@@ -333,3 +333,580 @@ private func getBsonValue(_ value: Any) -> BsonValue? {
+    /// - parameter work: The work to perform with the key in the path.
+    fileprivate func with<T>(pushedKey key: CodingKey, _ work: () throws -> T) rethrows -> T {
+        self.codingPath.append(key)
+        let ret: T = try work()
+        self.codingPath.removeLast()
+        return ret
+    }
+
+    /// Initializes `self` with the given top-level container and options.
+    fileprivate init(referencing container: BsonValue?, at codingPath: [CodingKey] = [], options: BsonDecoder._Options) {
+        self.storage = _BsonDecodingStorage()
+        self.storage.push(container: container)
+        self.codingPath = codingPath
+        self.options = options
+    }
+
+    // Returns the data stored in this decoder as represented in a container keyed by the given key type.
+    public func container<Key>(keyedBy type: Key.Type) throws -> KeyedDecodingContainer<Key> {
+        guard self.storage.topContainer != nil  else {
+            throw DecodingError.valueNotFound(KeyedDecodingContainer<Key>.self,
+                                  DecodingError.Context(codingPath: self.codingPath,
+                                                        debugDescription: "Cannot get keyed decoding container -- found null value instead."))
+        }
+
+        guard let topContainer = self.storage.topContainer as? Document else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: Document.self, reality: self.storage.topContainer)
+        }
+
+        let container = _BsonKeyedDecodingContainer<Key>(referencing: self, wrapping: topContainer)
+        return KeyedDecodingContainer(container)
+    }
+
+    // Returns the data stored in this decoder as represented in a container appropriate for holding a single primitive value.
+    public func singleValueContainer() throws -> SingleValueDecodingContainer {
+        return self
+    }
+
+    // Returns the data stored in this decoder as represented in a container appropriate for holding values with no keys.
+    public func unkeyedContainer() throws -> UnkeyedDecodingContainer {
+        guard self.storage.topContainer != nil else {
+            throw DecodingError.valueNotFound(UnkeyedDecodingContainer.self,
+                                  DecodingError.Context(codingPath: self.codingPath,
+                                                        debugDescription: "Cannot get unkeyed decoding container -- found null value instead."))
+        }
+
+        guard let arr = self.storage.topContainer as? [BsonValue?] else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: [BsonValue?].self, reality: self.storage.topContainer)
+        }
+
+        return _BsonUnkeyedDecodingContainer(referencing: self, wrapping: arr)
+    }
+}
+
+// Storage for a _BsonDecoder.
+private struct _BsonDecodingStorage {
+
+    /// The container stack, consisting of `BsonValue?`s. 
+    private(set) fileprivate var containers: [BsonValue?] = []
```

Comment:
one place I diverged from the `JSONEncoder` is using optionals/nils rather than `NSNull` to represent null values... this felt more intuitive to me because we can specify that if a container exists it's definitely a `BsonValue` rather than `Any`... but can you think of any advantage to using `NSNull` instead?